File: static/slides/2021/JuliaInPhysics2021/scripts/sudoku/sudoku.jl
# Simple Sudoku solver
# --------------------

function solve(puzzle)
    next(i, j) = i < 9 ? (i + 1, j) : (1, j + 1)
    function solveit(sol, pos)
        i, j = pos
        if j > 9
            return true
        elseif sol[i,j] > 0
            return solveit(sol, next(i, j))
        end
        m, n = (i - 1) ÷ 3, (j - 1) ÷ 3
        for k in 1:9
            if all(sol[i,j] != k for i in 1:9) &&
               all(sol[i,j] != k for j in 1:9) &&
               all(sol[i,j] != k for j in 3n+1:3n+3,
                                     i in 3m+1:3m+3)
                sol[i,j] = k
                solveit(sol, next(i, j)) && return true
            end
        end
        sol[i,j] = 0
        return false
    end
    sol = copy(puzzle)
    return solveit(sol, (1, 1)) ? sol : nothing
end

function load(file)
    puzzle = zeros(Int, 9, 9)
    i = 1
    for line in eachline(file)
        line = strip(line)
        (isempty(line) || startswith(line, '#')) && continue
        for (j, m) in enumerate(eachmatch(r"[0-9]", line))
            puzzle[i,j] = parse(Int, m.match)
        end
        i += 1
    end
    return puzzle
end

puz = load(stdin)
sol = solve(puz)
if sol === nothing
    println("no solution")
else
    for i in 1:9
        println(join(sol[i,:], ' '))
    end
end

File: scripts/sudoku/sudoku.py
# Simple Sudoku solver
# --------------------

def solve(puzzle):
    def next(i, j):
        if j < 8:
            return (i, j + 1)
        else:
            return (i + 1, 0)

    def solveit(sol, pos):
        i, j = pos
        if i > 8:
            return True
        elif sol[i][j] > 0:
            return solveit(sol, next(i, j))

        m, n = i // 3, j // 3
        for k in range(1, 10):
            ok = True
            for jj in range(9):
                if sol[i][jj] == k:
                    ok = False
                    break
            if not ok:
                continue
            for ii in range(9):
                if sol[ii][j] == k:
                    ok = False
                    break
            if not ok:
                continue
            for ii in range(3*m, 3*m+3):
                for jj in range(3*n, 3*n+3):
                    if sol[ii][jj] == k:
                        ok = False
                        break
                if not ok:
                    break
            if not ok:
                continue
            sol[i][j] = k
            if solveit(sol, next(i, j)):
                return True

        sol[i][j] = 0
        return False

    import copy
    sol = copy.deepcopy(puzzle)
    if solveit(sol, (0, 0)):
        return sol 
    else:
        return None

def load(file):
    import re
    puzzle = []
    for line in file:
        line = line.strip()
        if not line or line.startswith('#'):
            continue
        row = []
        for m in re.findall(r'[0-9]', line):
            row.append(int(m))
        puzzle.append(row)
    return puzzle

import sys
puz = load(sys.stdin)
sol = solve(puz)
if not sol:
    print("no solution")
else:
    for i in range(9):
        print(' '.join(map(str, sol[i])))